jax_md/smap.py

```diff
@@ -463,3 +463,120 @@ def fn_mapped(R, neighbor, **dynamic_kwargs):
+        upon reduction. This is supplied to np.sum and so the same convention
+        is used.
+    kwargs: Arguement providing parameters to the mapped function. In cases
+        where no species information is provided, these should either be 1)
+        a scalar, 2) an ndarray of shape [n] based on the central atom,
+        3) an ndarray of shape [n, n, n] defining triplet interactions.
+        If species information is provided, then the parameters should
+        be specified as either 1) a scalar,  2) an ndarray of shape
+        [max_species], 3) an ndarray of shape [max_species, max_species,
+        max_species] defining triplet interactions.
+
+  Returns:
+    A function fn_mapped.
+
+    If species is None or statically specified, then fn_mapped takes as
+    arguments an ndarray of positions of shape [n, spatial_dimension].
+
+    If species is Dynamic then fn_mapped takes as input an ndarray of shape
+    [n, spatial_dimension], an integer ndarray of species of shape [n], and
+    an integer specifying the maximum species.
+
+    The mapped function can also optionally take keyword arguments that get
+    threaded through the metric.
+  """
+  def extract_parameters_by_dim(kwargs, dim: Union[int, List[int]] = 0):
+    """Helper function that extract parameters from a dictionary via dimension.
+
+    Args:
+      kwargs: dictionary where values are all ndarrays
+      dim: dimension of parameters to extract
+
+    Returns:
+      A dictionary with the extracted elements of kwargs of the correct
+      dimension.
+    """
+    if isinstance(dim, int):
+      dim = [dim]
+    return {name: value for name, value in kwargs.items() if value.ndim in dim}
+
+  if species is None:
+    def fn_mapped(R, **dynamic_kwargs) -> Array:
+      d = space.map_product(partial(metric, **dynamic_kwargs))
+      _kwargs = merge_dicts(kwargs, dynamic_kwargs)
+      _kwargs = _kwargs_to_parameters(species, **_kwargs)
+      dR = d(R, R)
+      compute_triplet = partial(fn, **_kwargs)
+      output = vmap(vmap(vmap(compute_triplet, (None, 0)), (0, None)), 0)(dR, dR)
+      return high_precision_sum(output,
+                                axis=reduce_axis,
+                                keepdims=keepdims) / 2.
+  elif isinstance(species, jnp.ndarray):
+    def fn_mapped(R, **dynamic_kwargs):
+      d = partial(metric, **dynamic_kwargs)
+      idx = onp.tile(onp.arange(R.shape[0]), [R.shape[0], 1])
+      dR = vmap(vmap(d, (None, 0)))(R, R[idx])
+
+      _kwargs = merge_dicts(kwargs, dynamic_kwargs)
+
+      mapped_args = extract_parameters_by_dim(_kwargs, [1, 3])
+      mapped_args = {arg_name: jnp.take(
+          arg_value, species, axis=0) for arg_name, arg_value in mapped_args.items()}
+      # While we support 2 dimensional inputs, these often make less sense
+      # as the parameters do not depend on the central atom
+      unmapped_args = extract_parameters_by_dim(_kwargs, [0, 2])
```

Comment:
After a bit more discussion, can we for the moment throw an error if the dimension is 1 or 2? Both of these don't seem to have an obvious interpretation and so maybe better to punt.

Reply:
Done.